src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,195 @@
+                    await AddTextDocumentEdits(
+                        textDocumentEdits, applyChangesOperation, solution, changedDocuments,
+                        applyChangesOperation.ChangedSolution.GetDocument, solution.GetDocument,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Changed analyzer config documents
+                    // We won't get any results until https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+                    // is fixed.
+                    await AddTextDocumentEdits(
+                        textDocumentEdits, applyChangesOperation, solution, changedAnalyzerConfigDocuments,
+                        applyChangesOperation.ChangedSolution.GetAnalyzerConfigDocument, solution.GetAnalyzerConfigDocument,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Changed additional documents
+                    // We won't get any results until https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+                    // is fixed.
+                    await AddTextDocumentEdits(
+                        textDocumentEdits, applyChangesOperation, solution, changedAdditionalDocuments,
+                        applyChangesOperation.ChangedSolution.GetAdditionalDocument, solution.GetAdditionalDocument,
+                        cancellationToken).ConfigureAwait(false);
+                }
+
+                codeAction.Edit = new LSP.WorkspaceEdit { DocumentChanges = textDocumentEdits.ToArray() };
+            }
+
+            // Set up to run as command on the server instead of using WorkspaceEdits.
+            var commandOperations = operations.All(operation => !(operation is ApplyChangesOperation));
+            if (commandOperations || runAsCommand)
+            {
+                codeAction.Command = new LSP.Command
+                {
+                    CommandIdentifier = CodeActionsHandler.RunCodeActionCommandName,
+                    Title = codeAction.Title,
+                    Arguments = new object[] { data }
+                };
+            }
+
+            return codeAction;
+
+            // Local functions
+            static async Task AddTextDocumentEdits<T>(
+                ArrayBuilder<TextDocumentEdit> textDocumentEdits,
+                ApplyChangesOperation applyChangesOperation,
+                Solution solution,
+                IEnumerable<DocumentId> changedDocuments,
+                Func<DocumentId, T?> getNewDocumentFunc,
+                Func<DocumentId, T?> getOldDocumentFunc,
+                CancellationToken cancellationToken)
+                where T : TextDocument
+            {
+                foreach (var docId in changedDocuments)
+                {
+                    var newDoc = getNewDocumentFunc(docId);
+                    var oldDoc = getOldDocumentFunc(docId);
+
+                    var oldText = await oldDoc!.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
instead of hte `!` which will likel have hard to understand dumps, can we just do some Contract.ThrowIfNulls above here?